raiden/ui/web/src/app/components/token-network-selector/token-network-selector.component.ts

```diff
@@ -0,0 +1,149 @@
+import { Component, EventEmitter, forwardRef, OnInit, Output } from '@angular/core';
+import {
+    AbstractControl,
+    ControlValueAccessor,
+    FormControl, NG_VALIDATORS,
+    NG_VALUE_ACCESSOR,
+    ValidationErrors,
+    Validator,
+    ValidatorFn
+} from '@angular/forms';
+import { from, Observable } from 'rxjs';
+import { filter, flatMap, map, share, startWith, toArray } from 'rxjs/operators';
+import { UserToken } from '../../models/usertoken';
+import { RaidenService } from '../../services/raiden.service';
+import { amountToDecimal } from '../../utils/amount.converter';
+
+@Component({
+    selector: 'app-token-network-selector',
+    templateUrl: './token-network-selector.component.html',
+    styleUrls: ['./token-network-selector.component.css'],
+    providers: [
+        {
+            provide: NG_VALUE_ACCESSOR,
+            useExisting: forwardRef(() => TokenNetworkSelectorComponent),
+            multi: true
+        },
+        {
+            provide: NG_VALIDATORS,
+            useExisting: forwardRef(() => TokenNetworkSelectorComponent),
+            multi: true
+        }
+    ]
+})
+export class TokenNetworkSelectorComponent implements OnInit, ControlValueAccessor, Validator {
+
+    @Output() valueChanged = new EventEmitter<UserToken>();
+
+    private tokens$: Observable<UserToken[]>;
+    public filteredOptions$: Observable<UserToken[]>;
+
+    readonly tokenFc = new FormControl('', [this.addressValidatorFn(this.raidenService)]);
+
+    constructor(private raidenService: RaidenService) {
+    }
+
+    ngOnInit() {
+        this.tokens$ = this.raidenService.getTokens(true).pipe(
+            flatMap((tokens: UserToken[]) => from(tokens)),
+            toArray(),
+            map(value => value.sort(this._compareTokens)),
+            share()
+        );
+
+        this.filteredOptions$ = this.tokenFc.valueChanges.pipe(
+            startWith(''),
+            flatMap(value => this._filter(value))
+        );
+    }
+
+    // noinspection JSMethodCanBeStatic
+    trackByFn(token: UserToken): string {
+        return token.address;
+    }
+
+    tokenSelected(value: UserToken) {
+        this.tokenFc.setValue(value.address);
+    }
+
+    // noinspection JSMethodCanBeStatic
+    private _compareTokens(a: UserToken, b: UserToken): number {
+        if (a.connected === b.connected) {
+            return amountToDecimal(b.balance, b.decimals) - amountToDecimal(a.balance, a.decimals);
+        } else {
+            return a.connected ? -1 : 1;
+        }
+    }
+
+    private _filter(value?: string): Observable<UserToken[]> {
+        if (!value || typeof value !== 'string') {
+            return this.tokens$;
+        }
+
+        const keyword = value.toLowerCase();
+        return this.tokens$.pipe(
+            flatMap((tokens: UserToken[]) => from(tokens)),
```

Comment:
As you're dealing with an array of tokens, and filtering is a quite fast operation, I'd say you can save some ms/cicles by avoiding the `from/observableFilter/toArray` pattern and going straight to simple `tokens.filter`

Reply:
Still those java habits kicking. I will fix it :+1: